lib/utils/compose.js

```diff
@@ -0,0 +1,1087 @@
+		// Logger magically prefixes the log line with [Build] etc., but it doesn't
+		// work well with the spinner we're also showing. Manually build the prefix
+		// here and bypass the logger.
+		const prefix = getChalk().blue('[Build]') + '   ';
+
+		const offset = 10; // account for escape sequences inserted for colouring
+		this._prefixWidth =
+			offset + prefix.length + _.max(_.map(services, 'length'));
+		this._prefix = prefix;
+
+		// these are to handle window wrapping
+		this._maxLineWidth = null;
+		this._lineWidths = [];
+
+		this._startTime = null;
+		this._ended = false;
+		this._cancelled = false;
+		this._spinner = createSpinner();
+
+		this.streams = streams;
+	}
+
+	_handleEvent(service, event) {
+		this._serviceToDataMap[service] = event;
+	}
+
+	_handleInterrupt() {
+		this._cancelled = true;
+		this.end();
+		return process.exit(130); // 128 + SIGINT
+	}
+
+	start() {
+		process.on('SIGINT', this._handleInterrupt);
+		this._tty.hideCursor();
+		this._services.forEach(service => {
+			this.streams[service].write({ status: 'Preparing...' });
+		});
+		this._runloop = createRunLoop(this._display);
+		this._startTime = Date.now();
+	}
+
+	end(summary = null) {
+		if (this._ended) {
+			return;
+		}
+		this._ended = true;
+		process.removeListener('SIGINT', this._handleInterrupt);
+		this._runloop?.end();
+		this._runloop = null;
+
+		this._clear();
+		this._renderStatus(true);
+		this._renderSummary(summary ?? this._getServiceSummary());
+		this._tty.showCursor();
+	}
+
+	_display() {
+		this._clear();
+		this._renderStatus();
+		this._renderSummary(this._getServiceSummary());
+		this._tty.cursorUp(this._services.length + 1); // for status line
+	}
+
+	_clear() {
+		this._tty.deleteToEnd();
+		this._maxLineWidth = this._tty.currentWindowSize().width;
+	}
+
+	_getServiceSummary() {
+		const _ = require('lodash');
+
+		const services = this._services;
+		const serviceToDataMap = this._serviceToDataMap;
+
+		return _(services)
+			.map(function(service) {
+				const { status, progress, error } = serviceToDataMap[service] ?? {};
+				if (error) {
+					return `${error}`;
+				} else if (progress) {
+					const bar = renderProgressBar(progress, 20);
+					if (status) {
+						return `${bar} ${status}`;
+					}
+					return `${bar}`;
+				} else if (status) {
+					return `${status}`;
+				} else {
+					return 'Waiting...';
+				}
+			})
+			.map((data, index) => [services[index], data])
+			.fromPairs()
+			.value();
+	}
+
+	_renderStatus(end) {
+		if (end == null) {
+			end = false;
+		}
+		const moment = require('moment');
+		require('moment-duration-format')(moment);
+
+		this._tty.clearLine();
+		this._tty.write(this._prefix);
+		if (end && this._cancelled) {
+			this._tty.writeLine('Build cancelled');
+		} else if (end) {
+			const serviceCount = this._services.length;
+			const serviceStr =
+				serviceCount === 1 ? '1 service' : `${serviceCount} services`;
+			const durationStr = this._startTime == null ? 'unknown time' : moment
```

Comment:
This was changed to have the `this._startTime == null ? 'unknown time' :` check since the type for `this._startTime` is nullable